File: spring-java-starter/spring-java-mysql/starter/src/main/java/io/github/mrgsrylm/entity/BaseEntity.java
package io.github.mrgsrylm.entity;

import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import java.time.LocalDateTime;

@Data
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@MappedSuperclass
public abstract class BaseEntity {
    @Column(name = "CREATED_USER")
    protected String createdUser;

    @Column(name = "CREATED_AT")
    protected LocalDateTime createdAt;

    @Column(name = "UPDATED_USER")
    protected String updatedUser;

    @Column(name = "UPDATED_AT")
    protected LocalDateTime updatedAt;

   @PrePersist
   public void prePersist() {
       this.createdUser = getUsernameFromAuthentication();
       this.createdAt = LocalDateTime.now();
   }

   @PreUpdate
   public void preUpdate() {
       this.updatedUser = getUsernameFromAuthentication();
       this.updatedAt = LocalDateTime.now();
   }

   private String getUsernameFromAuthentication() {
       Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
       if (authentication != null && authentication.getPrincipal() instanceof CustomUserDetails userDetails) {
           return userDetails.getUsername();
       }
       return "anonymousUser";
   }
}

File: spring-kotlin-starter/spring-kotlin-mysql/starter/src/main/kotlin/io/github/mrgsrylm/entity/BaseEntity.kt
package io.github.mrgsrylm.entity

import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import java.time.LocalDateTime

@MappedSuperclass
abstract class BaseEntity {
    @Column(name = "CREATED_USER")
    lateinit var createdUser: String;

    @Column(name = "CREATED_AT")
    lateinit var createdAt: LocalDateTime;

    @Column(name = "UPDATED_USER")
    lateinit var updatedUser: String;

    @Column(name = "UPDATED_AT")
    lateinit var updatedAt: LocalDateTime;

    @PrePersist
    fun prePersist() {
        this.createdUser = getUsernameFromAuthentication()
        this.createdAt = LocalDateTime.now()
    }

    @PreUpdate
    fun preUpdate() {
        this.updatedUser = getUsernameFromAuthentication()
        this.updatedAt = LocalDateTime.now()
    }

    private fun getUsernameFromAuthentication(): String {
        return "anonymousUser";
    }
}